tests/integration/query.ts

```diff
@@ -50,3 +59,139 @@ it('universal autocomplete sets search intents', async () => {
   expect(statefulCore.state.query.searchIntents).toEqual(['NEAR_ME']);
 });
 
+
+describe('ensure correct results from latest request', () => {
+  jest.useFakeTimers();
+  const mockAutoCompleteFn = jest.fn(
+    async (request: VerticalAutocompleteRequest | UniversalAutocompleteRequest) => {
+      const waitTime = request.input?.length;
+      return new Promise(res => setTimeout(() => res({ results: [ {value: request.input} ]}), waitTime));
+    }
+  );
+
+  const mockedCore: any = {
+    verticalAutocomplete: mockAutoCompleteFn,
+    universalAutocomplete: mockAutoCompleteFn,
+    verticalSearch: jest.fn( async (request: VerticalSearchRequest) => {
+      const waitTime = request.query?.length;
+      return new Promise(res => setTimeout(() => res(
+        { verticalResults: { results: [request.query] } }), waitTime));
+    }),
+    universalSearch: jest.fn( async (request: UniversalSearchRequest) => {
+      const waitTime = request.query?.length;
+      return new Promise(res => setTimeout(() => res(
+        { verticalResults: [{ results: [request.query] }] }), waitTime));
+    })
+  };
+  const stateManager = new ReduxStateManager();
+  const httpManager = new HttpManager();
+  const statefulCore = new StatefulCore(mockedCore, stateManager, httpManager);
+  statefulCore.setVerticalKey('someKey');
+  const queries = ['really long request', 'short', 'long request'];
+  const updateResult = jest.fn();
+
+  beforeEach(() => {
+    jest.clearAllMocks();
+  });
+
+  it('vertical autocomplete get correct results based on up-to-date response', async () => {
+    statefulCore.addListener({
+      valueAccessor: state => state.vertical?.autoComplete?.results,
+      callback: updateResult
+    });
+
+    statefulCore.setQuery(queries[0]);
```

Comment:
nit but I feel like it's a little hard to read tests when they're written like this, since you have to mentally keep track of the what the value of the queries array is, and in this specific case also the length of each of the queries. Maybe we could have it so that the queries variable is a map of duration like "12 ms" to the query, or even just have the mock query be something like `"12"` and have the mock request parseInt that into the wait time

Reply:
oh I updated so that the request time is a bit clearer compare to just extracting it from the length of the queries. But now that I read your comment again, you are suggesting to pull the shared queries (and time) into the individual tests? I think that would lead to some redundancy no? lmk if the update might make it easier, or if you still prefer setting the queries and time hardcoded in each test instead of outside scrope